### PR TITLE
Use CMake for REx compilation

### PR DESCRIPTION
**Pull request by ekoeppen**
In order to facilitate adding new components to the ROM extension, build
the REx using CMake instead of make.

### CODE REVIEW

**Comment by MatthiasWM** [comment]
Nice! Looking forward to a working version!

**Comment by MatthiasWM** [comment]
This is truly amazing. I will try to compile my own Rex tonight to test your setup on Apple Silicon.

**Comment by ekoeppen** [comment]
> This is truly amazing. I will try to compile my own Rex tonight to test your setup on Apple Silicon.

It _should_ work :) As long as the DCL binaries and the NCT headers are in place, it's pretty straightforward.

One motivation is to make it a bit easier to add new code to the REx, not necessarily drivers, but other native modules. For the serial port updates, I'm working on a native module which enables the Einstein ports, it's quite easy to compile that now as as subproject of this setup. And the existing NSRuntime looks also like it could be compiled based on this.

Let me know how it goes, once you see it working, I'll merge unless there are some other findings.

File: Drivers/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,108 @@
+cmake_minimum_required(VERSION 3.10)
+enable_language(ASM)
+set(NCT_PROJECTS $ENV{NCT_PROJECTS})
+
+set(COMPILE_OPTIONS
+    -march=armv4 -mbig-endian
+    $<$<COMPILE_LANGUAGE:ASM>:
+        -I ${CMAKE_SOURCE_DIR}>
+    $<$<COMPILE_LANGUAGE:C,CXX>:
+        -O2
```
Comment (pguyot) [inline Drivers/CMakeLists.txt:10]: Are you replacing `-O3 -Os` with `-O2` here?

File: Drivers/cmake/arm-none-eabi.cmake [hunk 2]
```diff
@@ -0,0 +1,46 @@
+SET(CMAKE_SYSTEM_NAME Generic)
+set(CMAKE_SYSTEM_PROCESSOR arm)
+include(CMakeForceCompiler)
+set(TOOLCHAIN arm-none-eabi)
+if(NOT DEFINED TOOLCHAIN_PREFIX)
```
Comment (pguyot) [inline Drivers/cmake/arm-none-eabi.cmake:5]: Could we use find_program here instead to find arm-none-eabi-gcc? Provided the tool is on the path, it should work fine.

File: Drivers/cmake/arm-none-eabi.cmake [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        set(TOOLCHAIN_PREFIX "/usr/local")
+    elseif(CMAKE_HOST_SYSTEM_NAME STREQUAL Windows)
+        message(STATUS "Please specify the TOOLCHAIN_PREFIX !\n For example: -DTOOLCHAIN_PREFIX=\"C:/Program Files/GNU Tools ARM Embedded\" ")
+    else()
+        set(TOOLCHAIN_PREFIX "/usr")
+        message(STATUS "No TOOLCHAIN_PREFIX specified, using default: " ${TOOLCHAIN_PREFIX})
+    endif()
+endif()
+set(TOOLCHAIN_BIN_DIR ${TOOLCHAIN_PREFIX}/bin)
+set(TOOLCHAIN_INC_DIR ${TOOLCHAIN_PREFIX}/${TOOLCHAIN}/include)
```
Comment (pguyot) [inline Drivers/cmake/arm-none-eabi.cmake:18]: Where is this used?

**Review by pguyot — COMMENTED** [review]
Thank you! I understand you mean to replace make with cmake which is fine considering the main FLTK binary is built with cmake.

File: Drivers/CMakeLists.txt (hunk 1 above)
Comment (ekoeppen) [reply Drivers/CMakeLists.txt]: Good catch, I changed it back to what the Makefile uses.

File: Drivers/cmake/arm-none-eabi.cmake (hunk 3 above)
Comment (ekoeppen) [reply Drivers/cmake/arm-none-eabi.cmake]: Indeed, this doesn't seem to be used, I took it out now.

File: Drivers/cmake/arm-none-eabi.cmake (hunk 2 above)
Comment (ekoeppen) [reply Drivers/cmake/arm-none-eabi.cmake:5]: Sure! The toolchain setup for CMake cross compilation is a bit peculiar, but this makes sense to at least try.

**Comment by ekoeppen** [comment]
> Thank you! I understand you mean to replace make with cmake which is fine considering the main FLTK binary is built with cmake.

I'm hoping to actually use this as a starting point to compile other parts of the REx with just one setup, e.g. the NSRuntime native parts, or the serial port enabler I'm working on. CMake makes it quite simple to add new components to the overall project. Another side effect might be the ability to cross compile also on Windows, but that's at least for me not a goal right now.

File: Drivers/cmake/arm-none-eabi.cmake (hunk 2 above)
Comment (ekoeppen) [reply Drivers/cmake/arm-none-eabi.cmake]: Seems to work!

**Comment by ekoeppen** [comment]
> This is truly amazing. I will try to compile my own Rex tonight to test your setup on Apple Silicon.

I moved some things around to make this possibly more useful, hope it also works for you. The CI also produces the REx file as an artifact which could be tried too.

**Comment by MatthiasWM** [comment]
> Thank you! I understand you mean to replace make with cmake which is fine considering the main FLTK binary is built with cmake.

I don't thing that this is related to FLTK. CMake is great for generating build systems independently of the host platform. We use it with FLTK so we can compile our code on any of the three desktop platforms and even cross-compile for embedded systems. Current version of CMake can create dependencies on git archives which it downloads and builds for you. VisualStudio integrated CMake support a while ago, and the CLion IDE is built around CMake. It's becoming somewhat of a standard vs. autoconf/configure which never really ran on Windows. And you can generate Makefiles, or configurations for IDEs.

So when Eckhart is dome, I expect that we can write NewtonScript apps in Xcode and have them in the Rex?! :-)

**Comment by MatthiasWM** [comment]
This is how far I got on my M1 Macbook:
```
# In 'Einstein' directory checked out from GitHub:
export NCT_PROJECTS=/Users/matt/dev/Newton/NewtonDev/NewtonDev/Cpp/NCT_Projects/
brew tap ArmMbed/homebrew-formulae
brew install arm-none-eabi-gcc
git clone https://github.com/pguyot/DCL.git DCL.git
cd DCL.git/
cmake -S . -B build -DCMAKE_BUILD_TYPE=Debug
cmake --build build/
cd build && ctest . -C Debug
cmake --install build/
cd ..
cmake -S Drivers -B Drivers/build -DCMAKE_TOOLCHAIN_FILE=cmake/newton-cross.cmake
cmake --build Drivers/build
```
```
/Users/matt/dev/Einstein4iOS/Drivers/NewtonRuntime/NewtonRuntime.cpp:30:10: fatal error: NSandDDKIncludes.h: No such file or directory
   30 | #include <NSandDDKIncludes.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
```
```
find . -name NSandDDKIncludes.h
```
```
./Packages/iOS Transport/iOS Sharing_NCT/ObjCBridge Module/NSandDDKIncludes.h
```
I got quite far IMHO. How do you want me to continue?

**Comment by ekoeppen** [comment]
> export NCT_PROJECTS=/Users/matt/dev/Newton/NewtonDev/NewtonDev/Cpp/NCT_Projects/
> ``` 
> /Users/matt/dev/Einstein4iOS/Drivers/NewtonRuntime/NewtonRuntime.cpp:30:10: fatal error: NSandDDKIncludes.h: No such file or directory
>    30 | #include <NSandDDKIncludes.h>
>       |          ^~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> ```
> 
>  ```
> find . -name NSandDDKIncludes.h
> ```
> 
>  ```
> ./Packages/iOS Transport/iOS Sharing_NCT/ObjCBridge Module/NSandDDKIncludes.h
> ```

Not sure if NewtonDev is now up to date with what is needed for REx compilation. Before updating that and messing with StuffIt, maybe it's better to move to https://github.com/ekoeppen/NCT_Projects instead? Or then copy the headers from the iOS transport into the NewtonRuntime folder (I can add that as part of the PR)?

**Comment by MatthiasWM** [comment]
Whatever we need to change, I am just going by your `README.md`, so I would update that file instead of the PR.
```
MacBook-Pro:Einstein4iOS matt$ git clone https://github.com/ekoeppen/NCT_Projects.git NCT_Projects
MacBook-Pro:Einstein4iOS matt$ export NCT_PROJECTS=/Users/matt/dev/Einstein4iOS/NCT_Projects/
MacBook-Pro:Einstein4iOS matt$ cmake --build Drivers/build --target clean
MacBook-Pro:Einstein4iOS matt$ cmake --build Drivers/build 
[  1%] Building ASM object CMakeFiles/Runtime.dir/Glue.s.obj
[  3%] Building CXX object CMakeFiles/Runtime.dir/UGlue.cpp.obj
[  5%] Building CXX object CMakeFiles/Runtime.dir/NewtonRuntime/NewtonRuntime.cpp.obj
/Users/matt/dev/Einstein4iOS/Drivers/NewtonRuntime/NewtonRuntime.cpp:30:10: fatal error: NSandDDKIncludes.h: No such file or directory
   30 | #include <NSandDDKIncludes.h>
      |          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [CMakeFiles/Runtime.dir/NewtonRuntime/NewtonRuntime.cpp.obj] Error 1
make[1]: *** [CMakeFiles/Runtime.dir/all] Error 2
make: *** [all] Error 2
MacBook-Pro:Einstein4iOS matt$ find . -name NSandDDKIncludes.h
./Packages/iOS Transport/iOS Sharing_NCT/ObjCBridge Module/NSandDDKIncludes.h
./NCT_Projects/Includes/NSandDDKIncludes.h
MacBook-Pro:Einstein4iOS matt$ pwd
/Users/matt/dev/Einstein4iOS
```

**Comment by MatthiasWM** [comment]
OK, I had to configure again:
```
MacBook-Pro:Einstein4iOS matt$ cmake -S Drivers -B Drivers/build -DCMAKE_TOOLCHAIN_FILE=cmake/newton-cross.cmake
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_TOOLCHAIN_FILE


-- Build files have been written to: /Users/matt/dev/Einstein4iOS/Drivers/build
```
And then it compiled all the way through
```
MacBook-Pro:Einstein4iOS matt$ cmake --install Drivers/build 
-- Install configuration: ""
-- Installing: /Users/matt/dev/Einstein4iOS/Drivers/../_Data_/Einstein.rex
```
And I got my new Rex.

**Comment by MatthiasWM** [comment]
It should probably be
```
cmake -S Drivers -B Drivers/build -DCMAKE_TOOLCHAIN_FILE=Drivers/cmake/newton-cross.cmake
```
adding `Drivers/` to the path, but even then it does not pick up CMAKE_TOOLCHAIN_FILE.

I used a super old download of the driver tools from 10+ years ago. Using your repo solves it, and that's fine I guess.

**Comment by ekoeppen** [comment]
> It should probably be
> 
> ```
> cmake -S Drivers -B Drivers/build -DCMAKE_TOOLCHAIN_FILE=Drivers/cmake/newton-cross.cmake
> ```
> 
> adding `Drivers/` to the path, but even then it does not pick up CMAKE_TOOLCHAIN_FILE.

I noticed that the toolchain file needs to be referenced relative to the source directory (a bit confusing), so it is indeed `cmake/...` without `Drivers`. As an aside, it would be really great if CMake would handle mixed non-cross and cross compile projects, then we could call the drivers CMakeLists.txt directly from Einstein's CMakeLists.txt, but I haven't found any hint as to that being possible.

**Comment by MatthiasWM** [comment]
It works both ways for me. Maybe it's not used on macOS builds?

**Comment by ekoeppen** [comment]
> It works both ways for me. Maybe it's not used on macOS builds?

Interesting, here it fails when I do this from the Einstein root:

```
cmake -S Drivers -B Drivers/build -DCMAKE_TOOLCHAIN_FILE=Drivers/cmake/newton-cross.cmake
```

The documentation was updated for CMake 3.20, there it says that the path is resolved relative to the build or the source directory (I'm using 3.19 here), I've set the minimum CMake version to 3.10, and the CI seems to be happily doing the compilation with that, maybe leave it as is for now.

**Comment by MatthiasWM** [comment]
Ah, interesting. I am using CMake 3.22.0.

**Comment by MatthiasWM** [comment]
For later back reference:
```
git clone https://github.com/ekoeppen/NCT_Projects
export NCT_PROJECTS=/Users/matt/dev/NCT_Projects/
...
```

**Pull request merged**